consumer.go

```diff
@@ -162,8 +187,70 @@ func (c *Consumer) run() {
 			}
 
 		case <-c.done:
+			log.Println("Consumer initiating shutdown sequence")
+			// Send a CLS to all Cmd Channels for all connections
 			c.close()
+			log.Println("awaiting connection waitgroup")
+			// Wait for all runConn routines to return
 			c.join.Wait()
+			// At this point all runConn routines have returned, therefore we know
+			// we won't be receiving and new messages from nsqd servers. Now we can
+			// begin the processes of draining any in-flight messages and issuing a
+			// REQ command for each message.
+			log.Println("draining and requeueing remaining in-flight messages")
+			// drain and requeue any in-flight messages
+			c.drainRemaining()
+			log.Println("closing and cleaning up connections")
+			// Cleanup remaining connections
+			c.mtx.Lock()
+			for addr, cm := range c.conns {
+				delete(c.conns, addr)
+				// At this point we have drained all Messages from our main msgs channel and
+				// sent them REQ commands for each on their associated CmdChan. However, we can not simply just
+				// close the CmdChan for each connection yet. These channels are buffered and if
+				// we simply call closeCommand(cm.CmdChan) here there is a race, as the writeConn routines
+				// may not have finished processing all the REQ commands.
+				// Therefore we check the length of the channel and await for it to reach 0. If for some reason
+				// it fails to drain after a number of attempts we continue on and allow the messages to simply timeout
+				// and be reqeueued by the nsqd server.
+				attempts := 0
```

Comment:
Is the drain timeout intended to be a total timeout or per connection?  If there are a lot of connections, this could take a long time if you have lots of attempts.  Maybe spawn a go routine per connection?

Reply:
yes I mention this in the PR description, though you might have missed it, it's at the bottom.

> Finally of note we've added a new config option, DrainTimeout, which defaults to 1s. It is used as the max time to wait PER writeConn routine to process any remaining REQs. It is theoretically possible that if there were a lot of messages to REQ and a lot of down NSQ instances at the time of drain/requeue that this could introduce a long shutdown time (1s * nsqd connections). Perhaps we should consider adding parallelism here to address this potential limitation.

Reply:
Ha...RTFM.  Sorry about that.  I would definitely vote parallelism with a go routine per connection.